src/commands.cpp

```diff
@@ -1974,60 +1988,178 @@ void CommandGetUA::procresult()
 
         if (!(ptr = client->json.getvalue()) || !(end = strchr(ptr, '"')))
         {
-            return(client->app->getua_result(API_EINTERNAL));
+            client->app->getua_result(API_EINTERNAL);
+            return;
         }
 
         // if there's no avatar, the value is "none" (not Base64 encoded)
         if (attributename == "+a" && !strncmp(ptr, "none", 4))
         {
-            return(client->app->getua_result(API_ENOENT));
+            client->app->getua_result(API_ENOENT);
+            return;
         }
 
-        int l;
+        // convert from ASCII to binary the received data
+        unsigned int datalen;
         byte* data;
-        l = (end - ptr) / 4 * 3 + 3;
-        data = new byte[l];
-        l = Base64::atob(ptr, data, l);
+        datalen = (end - ptr) / 4 * 3 + 3;
+        data = new byte[datalen];
+        datalen = Base64::atob(ptr, data, datalen);
 
-        if (attributename == "*!lstint" || attributename == "*!authring")
+        bool missingKeypair = false;
+//        bool nonHistoric = (attributename.at(1) == '!');
+
+        // handle the attribute data depending on the scope
+        char scope = attributename.at(0);
+        switch (scope)
         {
-            string d;
-            d.assign((char*)data, l);
+        case '*':   // private
+
+            // decrypt the data and build the TLV records
+            tlvRecords = TLVstore::containerToTLVrecords(string((char*) data, datalen), &client->key);
+            if (!tlvRecords)
+            {
+                LOG_err << "Cannot extract TLV records for private attribute " << attributename;
+                client->app->getua_result(API_EINTERNAL);
+                delete [] data;
+                return;
+            }
 
-            // Is the data a multiple of the cipher blocksize, then we're using
-            // a zero IV.
-            if (l % client->key.BLOCKSIZE == 0)
+            if (attributename == "*keyring")
             {
-                if (!PaddedCBC::decrypt(&d, &client->key))
+                // Ed25519 not initialized yet
+                if (!client->signkey)
```

Comment:
This looks like it's only executed if the local value `signkey` is still missing. Shouldn't it be executed if the keyring contains a `signkey`? Similar with the `chatkey` below ...